pmlc/dialect/pxa/transforms/subgroups.cc

```diff
@@ -0,0 +1,248 @@
+// Copyright 2020 Intel Corporation
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Affine/IR/AffineValueMap.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Support/DebugStringHelper.h"
+
+#include "pmlc/dialect/pxa/analysis/strides.h"
+#include "pmlc/dialect/pxa/analysis/uses.h"
+#include "pmlc/dialect/pxa/ir/ops.h"
+#include "pmlc/dialect/pxa/transforms/cache.h"
+#include "pmlc/dialect/pxa/transforms/pass_detail.h"
+#include "pmlc/dialect/pxa/transforms/tile.h"
+#include "pmlc/dialect/pxa/transforms/tile_accumulate.h"
+#include "pmlc/dialect/pxa/transforms/vectorize.h"
+#include "pmlc/util/logging.h"
+
+using namespace mlir; // NOLINT
+
+namespace pmlc::dialect::pxa {
+
+namespace {
+
+using llvm::SmallVector;
+
+struct SubgroupPlan {
+  int64_t subgroupSize;
+  SmallVector<int64_t, 6> innerTile;
+  SmallVector<int64_t, 6> subgroupTile;
+};
+
+std::ostream &operator<<(std::ostream &os, const SubgroupPlan &plan) {
+  size_t sz = plan.innerTile.size();
+  os << plan.subgroupSize << ":[";
+  for (size_t i = 0; i < sz; i++) {
+    os << plan.innerTile[i];
+    if (i != sz - 1) {
+      os << " ";
+    }
+  }
+  os << "]:[";
+  for (size_t i = 0; i < sz; i++) {
+    os << plan.subgroupTile[i];
+    if (i != sz - 1) {
+      os << " ";
+    }
+  }
+  os << "]";
+  return os;
+}
+
+struct SubgroupParams {
+  SmallVector<int64_t, 4> subgroupSizes;
+  int64_t maxRegsPerThread;
+};
+
+struct SubgroupCostModel {
+  SubgroupCostModel(const SubgroupParams &params, AffineParallelOp op)
+      : params(params), op(op) {
+    bestCost = std::numeric_limits<double>::infinity();
+    // Verify that we have only 1 results
```

Comment:
Can you say why we care about 1 result (it's easy to see that the code checks for results != 1)?